apps/common/test/lexical/ast_test.exs

```diff
@@ -47,20 +47,77 @@ defmodule Lexical.AstTest do
       text = ~q[
         foo(bar do baz, bat|
       ]
+
       path = cursor_path(text)
       assert path == []
     end
   end
 
+  describe "path_at/2" do
+    defp path_at(text) do
+      {position, document} = pop_cursor(text, as: :document)
+      Ast.path_at(document, position)
+    end
+
+    test "returns an error if the cursor cannot be found in a node" do
+      code = ~q[
+        |
+        defmodule Foo do
+        end
+      ]
+
+      assert {:error, :not_found} = path_at(code)
+    end
+
+    test "returns an error if the AST cannot be parsed" do
+      code = ~q[
+        defmodule |Foo do
+      ]
+
+      assert {:error, {[line: 2, column: 1], "missing terminator: end" <> _, ""}} = path_at(code)
+    end
+
+    test "returns a path to the innermost node at position" do
+      code = ~q[
+        defmodule Foo do
+          def bar do
+            %{foo: |:ok}
+          end
+        end
+      ]
+
+      assert {:ok, [{:__block__, _, [:ok]} | _]} = path_at(code)
+    end
+
+    test "returns a path containing all ancestors" do
+      code = ~q[
+        defmodule Foo do
+          def |bar do
+            :ok
+          end
+        end
+      ]
+
+      assert {:ok,
+              [
+                {:bar, _, nil},
+                {:def, _, _},
+                {_, _},
+                [{_, _}],
+                {:defmodule, _, _}
+              ]} = path_at(code)
```

Comment:
So this function actually replaces the `__CURSOR__` of the `cursor_path` function's result with the exact token?

Reply:
Sort of, I think. The goal of `path_at` is to get the path to the innermost terminal node at a given position, where terminal node is essentially a node that doesn't have any (meaningful) children. To do this, `path_at` uses the entire AST and can give a more accurate path.

To illustrate the difference, I played around in `MIX_ENV=test iex -S mix` a bit:

```elixir
iex(4)> {pos, doc} = Lexical.Test.CursorSupport.pop_cursor(~q[
...(4)> defmodule MyModule do
...(4)>   def foo(bar) do
...(4)>     bar         
...(4)>     |> baz()
...(4)>     |> Other.|buzz()
...(4)>     |> Other.more()
...(4)>   end
...(4)> end
...(4)> ], as: :document)

iex(5)> Lexical.Ast.path_at(doc, pos)
{:ok,
 [
   {:., [line: 5, column: 13],
    [
      {:__aliases__, [last: [line: 5, column: 8], line: 5, column: 8], [:Other]},
      :buzz
    ]},
   {{:., [line: 5, column: 13],
     [{:__aliases__, [last: [line: 5, ...], line: 5, ...], [:Other]}, :buzz]},
    [closing: [line: 5, column: 19], line: 5, column: 14], []},
   {:|>, [newlines: 1, line: 5, column: 5],
    [
      {:|>, [newlines: 1, line: 4, ...], [{...}, ...]},
      {{:., [...], ...}, [closing: [...], ...], []}
    ]},
   {:|>, [newlines: 1, line: 6, column: 5],
    [{:|>, [newlines: 1, ...], [...]}, {{:., ...}, [...], ...}]},
   {{:__block__, [line: 2, column: 16], [:do]},
    {:|>, [newlines: 1, line: 6, ...], [{...}, ...]}},
   [{{:__block__, [line: 2, ...], [...]}, {:|>, [...], ...}}],
   {:def, [do: [line: 2, ...], end: [...], ...], [{...}, ...]},
   {{:__block__, [...], ...}, {:def, ...}},
   [{{...}, ...}],
   {:defmodule, ...}
 ]}

iex(6)> Lexical.Ast.cursor_path(doc, pos)  
[
  {:__cursor__, [closing: [line: 5, column: 25], line: 5, column: 14], []},
  {:|>, [newlines: 1, line: 5, column: 5],
   [
     {:|>, [newlines: 1, line: 4, ...], [{...}, ...]},
     {:__cursor__, [closing: [...], ...], []}
   ]},
  {{:__block__, [line: 2, column: 16], [:do]},
   {:|>, [newlines: 1, line: 5, column: 5], [{:|>, ...}, {...}]}},
  [
    {{:__block__, [line: 2, column: 16], [:do]},
     {:|>, [newlines: 1, ...], [...]}}
  ],
  {:def, [do: [line: 2, column: 16], end: [line: 5, ...], line: 2, ...],
   [{:foo, ...}, [...]]},
  {{:__block__, [line: 1, ...], [...]}, {:def, [...], ...}},
  [{{:__block__, ...}, {...}}],
  {:defmodule, [...], ...}
]
```

Since `cursor_path` only operates on the fragment before the cursor, it doesn't give any contextual information about the call that the cursor is in.